### PR TITLE
Extended documentation of driving functions, added new examples

### PR DESCRIPTION
**Pull request by fs446**
originated from Sascha
please double check in order to merge to master

### CODE REVIEW

File: sfs/mono/drivingfunction.py [hunk 1]
```diff
@@ -739,37 +1322,36 @@ def esa_edge_dipole_2d_line(omega, x0, xs, alpha=3/2*np.pi, Nc=None, c=None):
...
-    idx = (r <= r_s)
-    for m in np.arange(Nc):
-        nu = m*np.pi/alpha
-        f = 1/epsilon[m] * np.cos(nu*phi_s) * np.cos(nu*phi)
-        d[idx] = d[idx] + f[idx] * jn(nu, k*r[idx]) * hankel2(nu, k*r_s)
-        d[~idx] = d[~idx] + f[~idx] * jn(nu, k*r_s) * hankel2(nu, k*r[~idx])
+    if np.isscalar(xref):
+        a = np.linalg.norm(xref)/np.linalg.norm(xref-xs)
+    else:
+        a = np.linalg.norm(xref-x0, axis=1)/np.linalg.norm(xs-xref)
```
Comment (mgeier) [inline sfs/mono/drivingfunction.py]: It is kinda hard to see in this diff but you seem to have changed `xref-xs` to `xs-xref`.
I guess it doesn't matter if you calculate the norm of the resulting vector, but I was still wondering if this was intentional?

File: sfs/mono/drivingfunction.py [hunk 2]
```diff
@@ -739,37 +1322,36 @@ def esa_edge_dipole_2d_line(omega, x0, xs, alpha=3/2*np.pi, Nc=None, c=None):
...
 
-    return -1j*np.pi/alpha * d
+    d_2d, _, _ = esa_edge_2d_line(omega, x0, xs, alpha=alpha, Nc=Nc, c=c)
+    d = 1j*np.sqrt(a) * d_2d
+    selection = util.source_selection_all(len(x0))
```
Comment (mgeier) [inline sfs/mono/drivingfunction.py]: Why do you calculate the source selection a second time?

What about this:

```python
d_2d, selection, _ = esa_edge_2d_line(...)
d = ...
return d, selection, secondary_source_point(omega, c)
```

**Comment by mgeier** [comment]
For reference, this is the continuation of #104.

File: sfs/mono/drivingfunction.py [hunk 3]
```diff
@@ -282,7 +478,32 @@ def wfs_25d_focused(omega, x0, n0, xs, ns, xref=[0, 0, 0], c=None,
...
+    -------
+    float
+        Complex weight for given angular frequency.
+
+    Notes
+    -----
+    .. math::
+
+        H(\w) = \begin{cases}
+            \sqrt{\i \wc} & \text{for } \w <= \w_\text{alias} \\
```
Comment (mgeier) [inline sfs/mono/drivingfunction.py]: See https://github.com/sfstoolbox/sfs-python/pull/104#discussion_r264035527

File: sfs/mono/drivingfunction.py [hunk 4]
```diff
@@ -419,61 +765,236 @@ def nfchoa_25d_plane(omega, x0, r0, n=[0, 1, 0], max_order=None, c=None):
...
     k = util.wavenumber(omega, c)
     ds = x0 - xs
     r = np.linalg.norm(ds, axis=1)
-    return - 1j/2 * k * xs[1] / r * hankel2(1, k * r)
-
+    d = - 1j/2 * k * xs[1] / r * hankel2(1, k * r)
+    selection = util.source_selection_all(len(x0))
+    return d, selection, secondary_source_line(omega, c)
 
 def sdm_2d_plane(omega, x0, n0, n=[0, 1, 0], c=None):
```
Comment (mgeier) [inline sfs/mono/drivingfunction.py]: sfs/mono/drivingfunction.py:841:1: E302 expected 2 blank lines, found 1

File: sfs/mono/drivingfunction.py [hunk 5]
```diff
@@ -483,19 +1004,18 @@ def sdm_25d_point(omega, x0, n0, xs, xref=[0, 0, 0], c=None):
...
     ds = x0 - xs
     r = np.linalg.norm(ds, axis=1)
-    return 1/2 * 1j * k * np.sqrt(xref[1] / (xref[1] - xs[1])) * \
+    d = 1/2 * 1j * k * np.sqrt(xref[1] / (xref[1] - xs[1])) * \
         xs[1] / r * hankel2(1, k * r)
-
+    selection = util.source_selection_all(len(x0))
+    return d, selection, secondary_source_point(omega, c)
 
 def esa_edge_2d_plane(omega, x0, n=[0, 1, 0], alpha=3/2*np.pi, Nc=None,
```
Comment (mgeier) [inline sfs/mono/drivingfunction.py]: sfs/mono/drivingfunction.py:1012:1: E302 expected 2 blank lines, found 1

File: sfs/mono/drivingfunction.py [hunk 6]
```diff
@@ -543,18 +1076,18 @@ def esa_edge_2d_plane(omega, x0, n=[0, 1, 0], alpha=3/2*np.pi, Nc=None,
 
     d[phi > 0] = -d[phi > 0]
 
-    return 4*np.pi/alpha * d
+    selection = util.source_selection_all(len(x0))
+    return 4*np.pi/alpha * d, selection, secondary_source_line(omega, c)
+
 
 
 def esa_edge_dipole_2d_plane(omega, x0, n=[0, 1, 0], alpha=3/2*np.pi, Nc=None,
```
Comment (mgeier) [inline sfs/mono/drivingfunction.py]: sfs/mono/drivingfunction.py:1084:1: E303 too many blank lines (3)

File: sfs/mono/drivingfunction.py [hunk 7]
```diff
@@ -419,61 +765,236 @@ def nfchoa_25d_plane(omega, x0, r0, n=[0, 1, 0], max_order=None, c=None):
     return 2*1j / r0 * d, selection, secondary_source_point(omega, c)
 
 
+def _max_order_circular_harmonics(N, max_order):
```
Comment (mgeier) [inline sfs/mono/drivingfunction.py]: This function is dead code and should be removed.

The function now lives here:

https://github.com/sfstoolbox/sfs-python/blob/8bb6387a7078c3ce654cd3790fe41b808230098f/sfs/util.py#L621-L645

File: sfs/mono/drivingfunction.py (hunk 1 above)
Comment (fs446) [reply sfs/mono/drivingfunction.py]: was intentionally to check a broadcast issue, forgot to recover this, sorry. Fixed.

File: sfs/mono/drivingfunction.py (hunk 2 above)
Comment (fs446) [reply sfs/mono/drivingfunction.py]: Initially also thought, but not done then. Now that you also suggest, changed to it and even shortened it.

File: sfs/mono/drivingfunction.py (hunk 3 above)
Comment (fs446) [reply sfs/mono/drivingfunction.py]: missed that, thanks for the hint again, now fixed.

File: sfs/mono/drivingfunction.py (hunk 4 above)
Comment (fs446) [reply sfs/mono/drivingfunction.py]: fixed, sorry, pep8 check was not done yet.

File: sfs/mono/drivingfunction.py (hunk 7 above)
Comment (fs446) [reply sfs/mono/drivingfunction.py]: removed now, was not 100% sure about it.

Comment (mgeier) [reply sfs/mono/drivingfunction.py]: The tests pass, so it should be fine!

**Pull request merged**